flink-runtime/src/main/java/org/apache/flink/runtime/state/filemerging/SegmentFileStateHandle.java

```diff
@@ -62,19 +77,58 @@ public class SegmentFileStateHandle implements StreamStateHandle {
      * @param scope The state's scope, whether it is exclusive or shared.
      */
     public SegmentFileStateHandle(
-            Path filePath, long startPos, long stateSize, CheckpointedStateScope scope) {
+            Path directoryPath,
+            Path filePath,
+            long startPos,
+            long stateSize,
+            CheckpointedStateScope scope) {
         this.filePath = filePath;
         this.stateSize = stateSize;
         this.startPos = startPos;
         this.scope = scope;
+        this.directoryStateHandle =
+                DirectoryStreamStateHandle.forPathWithSize(
+                        new File(directoryPath.getPath()).toPath());
```

Comment:
Will `forPathWithSize` be inefficiency?

Reply:
👍, I've stored the `DirectoryStreamStateHandle` of each subtask in `FileMergingSnapshotManager`. Now, they are only created once when job start.